### PR TITLE
Move from minisign to ring

### PR DESCRIPTION
**Pull request by ipetr0v**
This change switches `minisign` library for `ring`.

Fixes https://github.com/project-oak/oak/issues/1385

# Checklist
- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](/cloudbuild.yaml)
  - [x] I have updated [documentation](/docs/) accordingly.

### CODE REVIEW

File: docs/development.md
Comment (daviddrysdale) [inline docs/development.md:288]: (for a future PR)

This text talks about how to generate a signature of a module, but I don't think there's any explanation of *why* you might want such a thing.   Could we add an explanation somewhere (probably in concepts.md, possibly adapted from your helpful [explanation on Slack](https://project-oak.slack.com/archives/CHE9E13C3/p1598369605051800)) of what signature labels mean, then reference it from here?

File: oak_loader/src/options.rs
Comment (daviddrysdale) [inline oak_loader/src/options.rs:236]: Could we get a `TODO(#999)` here please?

File: oak_runtime/src/node/wasm/mod.rs
Comment (daviddrysdale) [inline oak_runtime/src/node/wasm/mod.rs:852]: (pre-existing) Quick question to check my understanding: would we expect this secrecy label to normally use disjunctions rather than (as here) conjunctions, when disjunctions are available?  

AIUI this currently builds a label like `code:hash ∧ pubkey:A ∧ pubkey:B ∧ pubkey:C`, meaning that the client has to specify *all* of those tags.

Comment (tiziano88) [reply oak_runtime/src/node/wasm/mod.rs:852]: Note that this function computes the **privilege** of the node, i.e. the **maximum authority** that the node can use in declassifying (or endorsing) data.

This means that, for instance, a node with privilege `code:hash ∧ pubkey:A ∧ pubkey:B ∧ pubkey:C` may declassify **any of** those tags; so data labelled with a confidentiality label of `pubkey:A` (alone) may be declassified by this node, and so can data labelled with `pubkey:B`, etc.

Please do let me know if it is still unclear

File: examples/trusted_information_retrieval/signatures.toml
Comment (tiziano88) [inline examples/trusted_information_retrieval/signatures.toml:5]: Note that this key-value format now splits the information about the hash of the code to this file, while public key and signature are in the file pointed to by path. 

I think it is worth trying to think ahead how this will be used in the future, which I imagine will be a verifiable log or a simple database that ties together hash, public key, and signature of each module. e.g. if you imagine that as a mySQL table, it would be a table with three columns, so I think it makes sense to encode the three pieces of information in the same file to begin with.

WDYT?

Can be a follow-up PR, but I'm mostly curious if you think it will evolve in that direction, or do you have something else in mind @ipetr0v

File: oak_runtime/src/node/wasm/tests.rs
Comment (tiziano88) [inline oak_runtime/src/node/wasm/tests.rs:76]: I may be missing something, but it seems that this is loading a signature, not creating one?

File: oak_runtime/src/node/wasm/mod.rs
Comment (daviddrysdale) [reply oak_runtime/src/node/wasm/mod.rs:852]: Doh, it's a privilege not a label &ndash; apologies for not reading closely enough.

File: docs/development.md
Comment (ipetr0v) [reply docs/development.md:288]: I have a partial explanation here:
https://github.com/project-oak/oak/pull/1396

Not sure if it's enough

File: examples/trusted_information_retrieval/signatures.toml
Comment (ipetr0v) [reply examples/trusted_information_retrieval/signatures.toml:5]: I was thinking that it would look like this:
```toml
[signatures]

# database_proxy
3606a86d9fa01d223dc3cb092c1aceb9e3afe0ed9799ac5d5b214b1610209f8f = [
  { path = "examples/trusted_information_retrieval/database_proxy.sign" },
  { url = "https://oaksignatures.io/signature1.sign" },
  { url = "https://oaksignatures.io/signature2.sign" },
]

# some_other_module
3606a86d9fa01d223dc3cb092c1aceb9e3afe0ed9799ac5d5b214b1610209f8f = [
  { url = "https://oaksignatures.io/signature3.sign" },
]
```
So that it will just contain paths/URLs of signature files. And thus signatures will accessible just by a URL.

I'm also not sure if we will need a verifiable log for signatures (we may only need it for the Oak binary hash itself).

**Comment by tiziano88** [comment]
adding @rbehjati as reviewer to try and disseminate knowledge and spread review a bit more across the team :)

File: examples/trusted_information_retrieval/database_proxy.sign
Comment (daviddrysdale) [reply examples/trusted_information_retrieval/database_proxy.sign:4]: No, I mean a separate source for the public key.

In a non-Oak context, if I were to download some app from a website, and also got both a signature and a public key from the same website, then it's tempting to verify all three together.  However, doing so doesn't really tell me anything &ndash; anyone who takes over the website can modify all three:
 - create a hacked version of the app
 - create a new keypair and serve the public key
 - use the new private key to create a new signature over the hacked app, and serve that.

For the signature to mean anything, I need to trust the public key, which means I need to acquire it by some separate path, i.e. a PKI.

Comment (tiziano88) [reply examples/trusted_information_retrieval/database_proxy.sign:4]: @daviddrysdale I see your point, but **I think** it does not apply here, since this is handling signatures for pieces of code, but not making any assumptions about how someone else (dat owner) would label their data with something that refers to the public key corresponding to those signatures.

This is similar to signing an Android application, and the play store checking that the signature indeed matches a public key which is bundled together with it. How someone (an end user, most likely) would get to find out about and trust such public key in the first place, is a separate problem (and one that is yet unsolved for Oak, I would say).

For now, from the data owner point of view, we expect that they magically know what public keys to trust, and would just label data accordingly, then Oak would just respect those labels.

How to make public keys discoverable, tie them to higher level identities, make sure there are no split view attacks, etc. are problems that I hope will be somewhat solved / solvable by something like key transparency / keybase, or some other PKI. Speaking of which, do you think Tink would help with any of these? (in my mind it serves a somewhat similar purpose but on the server side, i.e. handling private key rotation and matching, but maybe there is something more that it may be used for).

I also think we should look a bit more closely at how similar mechanisms work on Android, iOS and macOS, since I think there are some similarities.

File: examples/trusted_information_retrieval/signatures.toml
Comment (tiziano88) [reply examples/trusted_information_retrieval/signatures.toml:5]: But why not make the hash part of the thing that is downloaded via the URL? To avoid fetching them upfront?

File: examples/certs/ed25519/README.md
Comment (tiziano88) [inline examples/certs/ed25519/README.md:1]: I don't remember if I commented on this before (cannot find it now, at least), but why are these under the existing `certs` folder? I would expect a `keys` folder to be more appropriate?

File: examples/trusted_information_retrieval/signatures.toml
Comment (tiziano88) [reply examples/trusted_information_retrieval/signatures.toml:5]: this is what I am suggesting:

```toml
signatures = [
  # database_proxy
  { path = "examples/trusted_information_retrieval/database_proxy.sign" },
  { url = "https://oaksignatures.io/signature1.sign" },
  { url = "https://oaksignatures.io/signature2.sign" },
  # some_other_module
  { url = "https://oaksignatures.io/signature3.sign" },
]
```

File: docs/development.md
Comment (ipetr0v) [reply docs/development.md:293]: After a discussion in https://github.com/project-oak/oak/issues/1066, we decided not to use signatures in the `truster_information_retrieval` example: https://github.com/project-oak/oak/issues/1407

**Comment by rbehjati** [comment]
I think the following line must be updated too:
https://github.com/project-oak/oak/blob/b8e35fae6a43dbb34cd0abc829ea979db43d6632/oak_abi/proto/label.proto#L72

File: docs/development.md
Comment (rbehjati) [reply docs/development.md:293]: Then perhaps add signing and a `signature.toml` file to the `hello_world` example (if there is not already a PR for that).

Comment (ipetr0v) [reply docs/development.md:293]: I'll repurpose the following PR (https://github.com/project-oak/oak/pull/1392) for this

File: oak_runtime/src/lib.rs
Comment (ipetr0v) [reply oak_runtime/src/lib.rs:1282]: Similar function is also defined in `oak_runtime`.
But currently we don't have an `oak_common` crate. Maybe we'll define it, but there is nothing much to be put there.

File: oak_loader/src/options.rs
Comment (ipetr0v) [reply oak_loader/src/options.rs:260]: I think in a case of a corrupted file it's better for the maintainer (a person who runs Oak) to just exclude it from the `signatures.toml`, because it's a critical error and it might be easy to miss it as an `error!` log.
And if a signature couldn't be loaded, Oak will work fine, but clients wouldn't be able to connect to Oak - and it's harder to check.

File: oak_runtime/src/lib.rs
Comment (rbehjati) [reply oak_runtime/src/lib.rs]: Do you mean there are other functions that are also used only in tests and in oak_loader?

**Review by rbehjati — APPROVED** [review]
LGTM!

File: oak_runtime/src/lib.rs
Comment (ipetr0v) [reply oak_runtime/src/lib.rs]: No, I think there are not, that's why I didn't think about creating an `oak_common` crate, since it will be too small.

File: oak_sign/src/main.rs
Comment (tiziano88) [inline oak_sign/src/main.rs:163]: Is it wise to generate a hash, store that in the signature file, but then calculate the actual signature on the original input bytes instead of the computed hash (which will presumably still sign the hash, though it may or may not be the same hash)? It seems it may make more sense to just sign the computed hash (even if it ends up being re-hashed as part of the signature)?

@benlaurie do you have an opinion on this?

**Review by tiziano88 — APPROVED** [review]

File: oak_sign/src/main.rs
Comment (ipetr0v) [reply oak_sign/src/main.rs:163]: [Ed25519](https://en.wikipedia.org/wiki/EdDSA#Ed25519) already uses SHA-512 for signing.
I think if we will start signing SHA-256 hashes with Ed25519 it may raise a lot of questions from the Oak application developers in the future.

Comment (ipetr0v) [reply oak_sign/src/main.rs:163]: > Is it wise to generate a hash, store that in the signature file, but then calculate the actual signature on the original input bytes instead of the computed hash

Hash that is stored in the file is used just as a reference, so when we want to check a Wasm file's signature, we compute its hash, and then find a corresponding file, that contains the same hash.

Comment (tiziano88) [reply oak_sign/src/main.rs:163]: Let's follow up separately, so we don't block this PR. Could you create a separate issue to track, so we don't forget?

Comment (ipetr0v) [reply oak_sign/src/main.rs:163]: Created an issue https://github.com/project-oak/oak/issues/1437

**Pull request merged**
